File: src/main/scala/com/datastax/home_spotify/exercises/Exercise1.scala
package com.datastax.home_spotify.exercises

import com.datastax.home_spotify.exercises.Constants.EXERCISE_1
import com.datastax.home_spotify.exercises.Schema._
import com.datastax.spark.connector._
import com.datastax.spark.connector.rdd.CassandraRDD
import org.apache.spark.rdd.RDD
import scala.collection.convert.Wrappers._

object Exercise1 extends BaseExercise {

  def main (args: Array[String]) {


    val sc = buildSparkContext(EXERCISE_1)

    /*
     * performers table structure
     *
     * CREATE TABLE IF NOT EXISTS performers (
     *   name TEXT,
     *   country TEXT,
     *   gender TEXT,
     *   type TEXT,
     *   born TEXT,
     *   died TEXT,
     *   styles LIST<TEXT>,
     *   PRIMARY KEY (name)
     * );
     */
    val rdd:CassandraRDD[CassandraRow] = sc.cassandraTable(KEYSPACE, PERFORMERS)

    val performerAndStyles:RDD[(String,String)] =
      rdd
        //TODO Transform a CassandraRow object into a (performer,list of styles) tuple
        //TODO Use the API CassandraRow.getString("???"), CassandraRow.getList[???]("???").toList ...
        .map[(String,List[String])](???)
        //TODO Extract all styles for each performer and create the tuple (style,performer)
        //TODO Each performer can have more than 1 style
        //TODO BE CAREFUL!!! The columns ordering ("style" and then "performer") is IMPORTANT
        .flatMap[(String,String)](???)

    //Save data back to Cassandra
    performerAndStyles.saveToCassandra(KEYSPACE,PERFORMERS_BY_STYLE,SomeColumns("style","performer"))
    
    sc.stop()
  }
}

File: src/main/java/com/datastax/home_spotify/exercises/Exercise1.java
package com.datastax.home_spotify.exercises;

import static com.datastax.home_spotify.exercises.Schema.*;

import com.datastax.spark.connector.japi.CassandraRow;
import com.datastax.home_spotify.entity.PerformerByStyle;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import scala.Tuple2;


import java.util.List;

import static com.datastax.spark.connector.japi.CassandraJavaUtil.javaFunctions;
import static com.datastax.spark.connector.japi.CassandraJavaUtil.mapToRow;
import static com.datastax.home_spotify.exercises.Constants.EXERCISE_1;
import static java.util.stream.Collectors.toList;

public class Exercise1 extends BaseExercise {

    public static void main(String[] args) {

        JavaSparkContext sc = buildSparkContext(EXERCISE_1);

        /*
         * performers table structure
         *
         * CREATE TABLE IF NOT EXISTS performers (
         *   name TEXT,
         *   country TEXT,
         *   gender TEXT,
         *   type TEXT,
         *   born TEXT,
         *   died TEXT,
         *   styles LIST<TEXT>,
         *   PRIMARY KEY (name)
         * );
         */
        final JavaRDD<CassandraRow> rows = javaFunctions(sc).cassandraTable(KEYSPACE, PERFORMERS);

        final JavaRDD<PerformerByStyle> performerAndStyles =
            rows
            //TODO Transform a CassandraRow object into a Tuple2<>(performer,list of styles)
            //TODO Use the API CassandraRow.getString("???")
            //TODO and CassandraRow.getList("???",CassandraJavaUtil.typeConverter(???.class)) ...
                    .map(row -> new Tuple2<String,List<String>>(null, null))
            //TODO Extract all styles for each performer using flatMap
            //TODO and create the POJO PerformerByStyle
                    .flatMap(tuple -> tuple._2()
                            .stream()
                            .map(style -> (PerformerByStyle) null)
                    .collect(toList()));

        // Save data back to Cassandra
        javaFunctions(performerAndStyles).writerBuilder(KEYSPACE, PERFORMERS_BY_STYLE, mapToRow(PerformerByStyle.class))
            .saveToCassandra();
    }

}